aiohomekit/zeroconf.py

```diff
@@ -131,39 +144,64 @@ def __init__(
     ):
         super().__init__(char_cache)
         self._async_zeroconf_instance = zeroconf_instance
+        self._waiters: dict[str, list[asyncio.Future]] = {}
 
     async def async_start(self):
         zc = self._async_zeroconf_instance.zeroconf
         if not zc:
             return self
 
-        # FIXME: Ideally we want to attach to the zeroconf instance here and monitor for
-        # 1. Hostname/port changes
-        # 2. C# getting incremented
-        # zc.async_add_listener(self, None)
-
-    async def async_stop(self):
-        # FIXME: Detach from zeroconf instance
-        pass
-
-    async def async_find(self, device_id: str) -> ZeroconfDiscovery:
-        device_id = device_id.lower()
-
-        async for device in self.async_discover():
-            if device.description.id == device_id:
-                return device
-
-        raise AccessoryNotFoundError(f"Accessory with device id {device_id} not found")
+        self._browser = find_brower_for_hap_type(
+            self._async_zeroconf_instance, self.hap_type
+        )
+        self._browser.service_state_changed.register_handler(self._handle_service)
 
-    async def async_discover(self) -> AsyncIterable[ZeroconfDiscovery]:
-        zc = self._async_zeroconf_instance.zeroconf
         infos = [
             AsyncServiceInfo(self.hap_type, record.alias)
             for record in zc.cache.get_all_by_details(self.hap_type, TYPE_PTR, CLASS_IN)
         ]
 
         await asyncio.gather(*(self._async_handle_service(info) for info in infos))
 
+        return self
+
+    def _handle_service(
+        self,
+        zeroconf: Zeroconf,
+        service_type: str,
+        name: str,
+        state_change: ServiceStateChange,
+    ):
+        if service_type != self.hap_type:
+            return
+        info = AsyncServiceInfo(service_type, name)
+        async_create_task(self._async_handle_service(info))
```

Comment:
We could end up creating a lot of tasks here as we could get a `ServiceStateChange.Added` then a `ServiceStateChange.Updated`

Also need to check how `ServiceStateChange.Removed` is handled

Reply:
Maybe a 0.5s debouncer here that always runs again after the cooldown of _TIMEOUT_MS + 1s?

Reply:
I haven't found a good Debouncer lib.  Maybe we should pull the one out of HASS and publish it as a PyPi

Reply:
Good point about Removed. Could either send a `None` description or ignore it. Ignore it easiest option to start with. Don't think we should do anything pro-active about it - if its just a bad network we might still have a somewhat stable TCP connection in flight.

Reply:
Or... we won't update a pairing's description, but we'd probably remove it from `self.discoveries`

Reply:
I think i'm going to pull in the HA one into aiohomekit and get it working, then pull it out into a seperate package later. I'll burn all the time i have this evening setting up a new thing, and won't be hard to do later, just add a dep and change an import.

Reply:
https://github.com/Jc2k/aiohomekit/pull/151/commits/3fc7783ee81ed5d89726801d204b8e0065466a66 pulls out the debouncer, and adds a async_trigger that runs in event loop but in sync context - reduces number of tasks created.

Reply:
Merged in main, doing live testing now